docs: Add lambda extension docs (#2809)
grafana/pyroscope: docs/sources/configure-client/aws-lambda.md

---
title: "AWS Lambda Profiling Extension"
menuTitle: "AWS Lambda Profiling Extension"
description: "Profiling AWS Lambda functions with Pyroscope"
weight: 100
---

> # AWS Lambda Profiling Extension

Reviewer: Because we use sentence case for headings, I've changed the capitalization of this heading.
```suggestion
# AWS Lambda profiling extension
```

> ## Introduction to Pyroscope Lambda Extension

Reviewer: We don't need the heading since this sentence acts as the introduction. It's also good practice to have a sentence between headings. 
```suggestion
```

> ## Introduction to Pyroscope Lambda Extension
> 
> The Pyroscope AWS Lambda Extension is a robust tool for profiling AWS Lambda functions, ensuring minimal latency impact. This profiling is essential for optimizing your functions.

Reviewer: We don't need the heading since this sentence acts as the introduction. It's also good practice to have a sentence between headings.

```suggestion
The Pyroscope AWS Lambda Extension is a robust tool for profiling AWS Lambda functions, ensuring minimal latency impact. This profiling is essential for optimizing your functions.
```

## Why Profile AWS Lambda Functions?

> AWS Lambda functions, while powerful and flexible, can lead to significant costs if not managed efficiently. One of the primary challenges with serverless architectures like AWS Lambda is that they can mask performance issues. Since Lambda functions are billed based on execution time and allocated memory, inefficiencies in code can lead to higher costs. Often, these costs accumulate unnoticed because of the following reasons:
> 
> 1. **Granular Billing**: Lambda functions are billed in milliseconds, which can make small inefficiencies seem insignificant at first. However, when scaled to thousands or millions of invocations, these inefficiencies can lead to substantial costs.
> 
> 2. **Complex Performance Profile**: Lambda functions may interact with various services and resources, making it challenging to pinpoint performance bottlenecks.
> 
> 3. **Variable Load**: The serverless nature of AWS Lambda means that functions might handle variable loads at different times, making it hard to optimize for every scenario.
> 
> Profiling Lambda functions helps in identifying these hidden performance bottlenecks, enabling developers to optimize their code for both performance and cost. Effective profiling can reveal inefficient code paths, unnecessary memory usage, and areas where the execution time can be reduced. By addressing these issues, organizations can significantly reduce their AWS bill, improve application responsiveness, and ensure a more efficient use of resources.

Reviewer: Excellent information! The suggestions below tighten the text and put one sentence per line. What do you think? 

```suggestion
AWS Lambda functions, while powerful and flexible, can lead to significant costs if not managed efficiently.
Serverless architectures, like AWS Lambda, can mask performance issues.
Since Lambda functions are billed based on execution time and allocated memory, code inefficiencies can lead to higher costs. Often, these costs accumulate unnoticed because of the following reasons:

* **Granular Billing**: Lambda functions are billed in milliseconds, which can make small inefficiencies seem insignificant at first. However, when scaled to thousands or millions of invocations, these inefficiencies can lead to substantial costs.

* **Complex Performance Profile**: Lambda functions may interact with various services and resources, making it challenging to pinpoint performance bottlenecks.

* **Variable Load**: The serverless nature of AWS Lambda means that functions might handle variable loads at different times, making it hard to optimize for every scenario.

Profiling Lambda functions helps identify these hidden performance bottlenecks, enabling developers to optimize their code for both performance and cost.
Effective profiling can reveal inefficient code paths, unnecessary memory usage, and areas where the execution time can be reduced.
By addressing these issues, organizations can significantly reduce their AWS bill, improve application responsiveness, and ensure a more efficient use of resources.
```

## How the Pyroscope Lambda Extension Works

This extension runs a relay server on the same network namespace as the Lambda function, ensuring minimal added latency.

![Lambda Extension Architecture](https://user-images.githubusercontent.com/23323466/186037668-44de7caa-6576-422a-b3f7-8416325f4a98.png)

For more details, refer to the [Building Extensions for AWS Lambda blog post](https://aws.amazon.com/blogs/compute/building-extensions-for-aws-lambda-in-preview/).

## Setting Up the Pyroscope Lambda Extension

### Step 1: Configure Your Lambda Function

Configure your Lambda function to use the extension. Find the latest release on our [releases page](https://github.com/grafana/pyroscope-lambda-extension/releases).

> ### Step 2: Environment Variable Setup
> 
> Configure the extension with the following environment variables:

Reviewer: ```suggestion
### Set up the environment variables 

Configure the extension with the following environment variables:
```

| `PYROSCOPE_REMOTE_ADDRESS`     | `https://profiles-prod-001.grafana.net` | Destination for relayed Pyroscope data       |
| `PYROSCOPE_BASIC_AUTH_USER`    | `""`                                    | HTTP Basic authentication user               |
| `PYROSCOPE_BASIC_AUTH_PASSWORD`| `""`                                    | HTTP Basic authentication password           |
| `PYROSCOPE_AUTH_TOKEN`         | `""`                                    | Authorization key (token authentication)     |
| `PYROSCOPE_SELF_PROFILING`     | `false`                                 | Whether to profile the extension itself      |
| `PYROSCOPE_LOG_LEVEL`          | `info`                                  | Log level (`error`, `info`, `debug`, `trace`)|
| `PYROSCOPE_TIMEOUT`            | `10s`                                   | HTTP client timeout (in Go duration format)  |
| `PYROSCOPE_NUM_WORKERS`        | `5`                                     | Number of relay workers                      |
| `PYROSCOPE_TENANT_ID`          | `""`                                    | Pyroscope tenant ID (for multi-tenancy)      |

> ### Step 3: Integrate Pyroscope SDK

Reviewer: ```suggestion
### Integrate the Pyroscope SDK
```

For a Golang Lambda function, integrate the Pyroscope SDK as follows:

```go
func HandleRequest(ctx context.Context) (string, error) {
    return "Hello world!", nil
}

func main() {
    pyroscope.Start(pyroscope.Config{
        ApplicationName: "simple.golang.lambda",
        ServerAddress:   "http://localhost:4040",
    })
    lambda.Start(HandleRequest)
}
```

Replace `simple.golang.lambda` with your application name. The `ServerAddress` must be `http://localhost:4040`.

> ## Use Cases
> 
> Once set up, you can use the Pyroscope UI to analyze your Lambda function's data, facilitating performance optimizations. For more on this, visit our [Pyroscope AWS Lambda Extension blog post](http://pyroscope.io/blog/profile-aws-lambda-functions).

Reviewer: ```suggestion
## Use cases

Once set up, you can use the Pyroscope UI to analyze your Lambda function's data to facilitate performance optimizations. For more on this, visit our [Pyroscope AWS Lambda Extension blog post](http://pyroscope.io/blog/profile-aws-lambda-functions).
```

> ## Sending Data to Pyroscope

Reviewer: ```suggestion
## Send data to Pyroscope
```

## Architecture

This extension runs a relay server on the same network namespace as the Lambda function, ensuring minimal added latency.

![Lambda Extension Architecture](https://user-images.githubusercontent.com/23323466/186037668-44de7caa-6576-422a-b3f7-8416325f4a98.png)

For more details, refer to the [Building Extensions for AWS Lambda blog post](https://aws.amazon.com/blogs/compute/building-extensions-for-aws-lambda-in-preview/).

## Set up the Pyroscope Lambda extension

To set up the Pyroscope Lamnda extension, you need to: 

1. Configure your Lamda function
1. Set up your environment variables 
1. Integrate the Pyroscope SDK

### Configure your Lambda function

Configure your Lambda function to use the extension. Find the latest release on our [releases page](https://github.com/grafana/pyroscope-lambda-extension/releases).

### Set up the environment variables 

Configure the extension with the following environment variables:

| Environment Variable           | Default Value                           | Description                                  |
| ------------------------------ | --------------------------------------- | -------------------------------------------- |

> | `PYROSCOPE_REMOTE_ADDRESS`     | `https://profiles-prod-001.grafana.net` | Destination for relayed Pyroscope data       |

Reviewer 2: The current default value is `https://ingest.pyroscope.cloud` – I'll update.

| `PYROSCOPE_BASIC_AUTH_USER`    | `""`                                    | HTTP Basic authentication user               |
| `PYROSCOPE_BASIC_AUTH_PASSWORD`| `""`                                    | HTTP Basic authentication password           |

> | `PYROSCOPE_AUTH_TOKEN`         | `""`                                    | Authorization key (token authentication)     |

Reviewer 2: There are 4 from 9 options that control authentication and authorisation:
 - `PYROSCOPE_BASIC_AUTH_USER` and `PYROSCOPE_BASIC_AUTH_PASSWORD` required for cloud.
 - `PYROSCOPE_TENANT_ID` might be needed for on-premises deployment.
 - `PYROSCOPE_AUTH_TOKEN` was required for OG cloud, and might still be useful if pyroscope is deployed on-premises and has some custom authentication layer that uses bearer token.

I think we could remove the latter from this page to avoid overloading users with rarely useful information.

```suggestion
```

| `PYROSCOPE_SELF_PROFILING`     | `false`                                 | Whether to profile the extension itself      |
| `PYROSCOPE_LOG_LEVEL`          | `info`                                  | Log level (`error`, `info`, `debug`, `trace`)|
| `PYROSCOPE_TIMEOUT`            | `10s`                                   | HTTP client timeout (in Go duration format)  |
| `PYROSCOPE_NUM_WORKERS`        | `5`                                     | Number of relay workers                      |

> | `PYROSCOPE_TENANT_ID`          | `""`                                    | Pyroscope tenant ID (for multi-tenancy)      |

Reviewer 2: Better to move it up closer to auth-related stuff
